### ISSUE TITLE
Networkd bond wireless inteface not setting DHCP configuration

### ISSUE THREAD

**Issue by blatte** [systemd/systemd#29147]
### systemd version the issue has been seen with

systemd 254.1-1

### Used distribution

Archlinux

### Linux kernel version used

6.4.12-arch1-1

### CPU architectures issue was seen on

None

### Component

systemd-networkd

### Expected behaviour you didn't see

Bond interface uses a wifi interface to connect.

### Unexpected behaviour you saw

Bond interface has connectivity with ethernet interface but not with wifi interface.

### Steps to reproduce the problem

When creating a bond network inteface with a cable interface (as primary slave) and a wireless interface, there is only connectivity with the cable interface. When switching (or starting) on wifi, nothing works. Can't even ping the gateway. It seems the ip and routes are not set for the wireless interface. I'm using iwd only to connect to the access point (EnableNetworkConfiguration=false), networkd runs the DHCP configuration.

Here is my configuration:

25-bond0.network
```
[Match]
Name=bond0

[Network]
DHCP=yes

[DHCPv4]
UseDomains=true

[DHCPv6]
UseDomains=true

[IPv6AcceptRA]
UseDomains=true
```

30-bond0.netdev
```
[NetDev]
Name=bond0
Kind=bond

[Bond]
Mode=active-backup
PrimaryReselectPolicy=always
MIIMonitorSec=1s
```

30-ethernet-bond0.network

```
[Match]
Name=ethusb0

[Network]
Bond=bond0
PrimarySlave=true
```

30-wifi-bond0.network

```
[Match]
Name=wlan0

[Network]
Bond=bond0
IgnoreCarrierLoss=3s
```

With the eth interface up all is working (with traffic indeed going through the bond and the eth intefraces) but when eth is down, the bond switches to wifi but there is a loss of all connectivity. When plugging eth back and all is working.

networkctl
```
IDX LINK    TYPE     OPERATIONAL SETUP
  1 lo      loopback carrier     unmanaged
  3 wlan0   wlan     enslaved    configured
  6 bond0   bond     routable    configured
  7 ethusb0 ether    no-carrier  configured
```
When starting with only wifi, the bond0 interface has no IP set. When plugging eth0, bond0 gets an IP. Unplugging eth again leaves the IP on the bond interface but, again, there is not connectivity.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
systemd[1]: Starting Network Configuration...
systemd-networkd[4633]: lo: Link UP
systemd-networkd[4633]: lo: Gained carrier
systemd-networkd[4633]: wlan0: Link UP
systemd-networkd[4633]: wlan0: Gained carrier
systemd-networkd[4633]: bond0: netdev ready
systemd-networkd[4633]: bond0: Link UP
systemd-networkd[4633]: bond0: Gained carrier
systemd-networkd[4633]: ethusb0: Link UP
systemd-networkd[4633]: bond0: Gained IPv6LL
systemd-networkd[4633]: wlan0: Connected WiFi access point: myap (....)
systemd-networkd[4633]: Enumeration completed
systemd[1]: Started Network Configuration.
systemd-networkd[4633]: bond0: netdev exists, using existing without changing its parameters
systemd-networkd[4633]: wlan0: Configuring with /etc/systemd/network/30-wifi-bond0.network.
systemd-networkd[4633]: bond0: Configuring with /etc/systemd/network/25-bond0.network.
systemd-networkd[4633]: ethusb0: Configuring with /etc/systemd/network/30-ethernet-bond0.network.
systemd-networkd[4633]: wlan0: Link DOWN
systemd-networkd[4633]: wlan0: Lost carrier
systemd-networkd[4633]: ethusb0: Link DOWN
systemd-networkd[4633]: wlan0: Link UP
systemd-networkd[4633]: ethusb0: Link UP
systemd-networkd[4633]: bond0: Lost carrier
systemd-networkd[4633]: bond0: DHCPv6 lease lost
systemd-networkd[4633]: wlan0: Connected WiFi access point: myap (....)
systemd-networkd[4633]: wlan0: Gained carrier
systemd-networkd[4633]: bond0: Gained carrier
```

**Comment by yuwata** [comment]
Please provide the debugging logs of networkd.

**Comment by blatte** [comment]
Here is a [debug log](https://gist.github.com/blatte/203445cf6e9257d14091e2568507cfca)

**Comment by yuwata** [comment]
Sorry for late response.
So, from the log, the bond interface receives RAs, but does not receive any DHCPv4 and DHCPv6 replies...

Let me confirm that each interface (without bonding) can receive DHCP replies, right?

**Comment by blatte** [comment]
That is correct. Both interfaces work fine separately and can get an address from the DHCP. When bonded however, the wifi interface gets no IP, but ethernet does.

**Comment by r-vdp** [comment]
@yuwata I am having pretty much exactly the same setup as OP and exactly the same issue, no connectivity when connected over wifi.

I am running systemd 257.5.

Do you have any idea on what could be the issue here?

**Comment by r-vdp** [comment]
From doing some tcpdump'ing, I can see the discover request when I listen in on the bond interface, and on the enslaved wifi interface, but I don't see them on the dhcp server, and not even on the wifi access point. So it seems that the packets either never actually get sent out by my machine (something in the network stack is dropping them, because of the bond's mac address being different from the one of the wifi interface, maybe?), or they get dropped immediately by the AP without even being captured by tcpdump (also maybe related to the mac addresses?).

My AP is running openwrt, the dhcp server is a linux box with systemd-networkd as the dhcp server.

Let me know if I can provide any additional info.

**Comment by arvidjaar** [comment]
> because of the bond's mac address being different from the one of the wifi interface, maybe?

Yes. It is the same problem as with bridging - MAC address in Ethernet packets must be the same as used during association with AP. Try using wireless MAC. Does it work?

**Comment by blatte** [comment]
Late response, apologies, but I found what was wrong after a looooooooong time: MAC  address spoofing. You cannot obfuscate your mac address (which was my default for my interfaces when not using a bond). Once I removed it, it finally worked. I'm afraid it is still an issue for me, but at least it works...
